wicketd/src/installinator_progress.rs

```diff
@@ -0,0 +1,277 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Tracker for installinator progress reports.
+//!
+//! This connects up the wicketd artifact server, which receives reports, to the
+//! update tracker.
+
+use std::{collections::HashMap, sync::Arc};
+
+use installinator_artifactd::ProgressReportStatus;
+use installinator_common::{CompletionEventKind, ProgressReport};
+use tokio::sync::{mpsc, oneshot, Mutex};
+use uuid::Uuid;
+
+/// Creates an [`IprManager`] and the artifact server and update tracker's interfaces to it.
+pub(crate) fn new(log: &slog::Logger) -> (IprArtifactServer, IprUpdateTracker) {
+    let running_updates = Arc::new(Mutex::new(HashMap::new()));
+
+    let ipr_artifact = IprArtifactServer {
+        log: log.new(
+            slog::o!("component" => "installinator progress reports, artifact server"),
+        ),
+        running_updates: running_updates.clone(),
+    };
+    let ipr_update_tracker = IprUpdateTracker {
+        log: log.new(
+            slog::o!("component" => "installinator progress reports, update tracker"),
+        ),
+        running_updates,
+    };
+
+    (ipr_artifact, ipr_update_tracker)
+}
+
+/// The artifact server's interface to the installinator tracker.
+#[derive(Debug)]
+#[must_use]
+pub(crate) struct IprArtifactServer {
+    log: slog::Logger,
+    running_updates: Arc<Mutex<HashMap<Uuid, RunningUpdate>>>,
+}
+
+impl IprArtifactServer {
+    pub(crate) async fn report_progress(
+        &self,
+        update_id: Uuid,
+        report: ProgressReport,
+    ) -> ProgressReportStatus {
+        let mut running_updates = self.running_updates.lock().await;
+        if let Some(update) = running_updates.get_mut(&update_id) {
+            slog::debug!(
+                self.log,
+                "progress report seen ({} completion events, {} progress events)",
+                report.completion_events.len(),
+                report.progress_events.len();
+                "update_id" => %update_id
+            );
+            match update.take() {
```

Comment:
Might be worth a brief comment that `.take()` leaves `update` in the `Invalid` state, and each arm _must_ set `update` back to a valid state. This is my least favorite part of enum type states; I usually end up with `Option<State>` (which is basically equivalent) and it's always unsatisfying.

Reply:
Done